consensus/src/marshal/ingress/handler.rs

```diff
@@ -88,3 +94,395 @@ impl<K: Array> Producer for Handler<K> {
+                *theirs > *mine
+            }
+            (Self::Finalized { .. }, _) => true,
+            (Self::Notarized { view: mine }, Self::Notarized { view: theirs }) => *theirs > *mine,
+            (Self::Notarized { .. }, _) => true,
+        }
+    }
+}
+
+impl<B: Block> Write for Request<B> {
+    fn write(&self, buf: &mut impl BufMut) {
+        match self {
+            Self::Block(commitment) => {
+                0u8.write(buf);
+                commitment.write(buf)
+            }
+            Self::Finalized { height } => {
+                1u8.write(buf);
+                height.write(buf)
+            }
+            Self::Notarized { view } => {
+                2u8.write(buf);
+                view.write(buf)
+            }
+        }
+    }
+}
+
+impl<B: Block> Read for Request<B> {
+    type Cfg = ();
+
+    fn read_cfg(buf: &mut impl Buf, _: &()) -> Result<Self, CodecError> {
+        let request = match u8::read(buf)? {
+            0 => Self::Block(B::Commitment::read(buf)?),
+            1 => Self::Finalized {
+                height: u64::read(buf)?,
+            },
+            2 => Self::Notarized {
+                view: u64::read(buf)?,
+            },
+            i => return Err(CodecError::InvalidEnum(i)),
+        };
+        Ok(request)
+    }
+}
+
+impl<B: Block> EncodeSize for Request<B> {
+    fn encode_size(&self) -> usize {
+        1 + match self {
+            Self::Block(block) => block.encode_size(),
+            Self::Finalized { height } => height.encode_size(),
+            Self::Notarized { view } => view.encode_size(),
+        }
+    }
+}
+
+impl<B: Block> Span for Request<B> {}
+
+impl<B: Block> PartialEq for Request<B> {
+    fn eq(&self, other: &Self) -> bool {
+        match (&self, &other) {
+            (Self::Block(a), Self::Block(b)) => a == b,
+            (Self::Finalized { height: a }, Self::Finalized { height: b }) => a == b,
+            (Self::Notarized { view: a }, Self::Notarized { view: b }) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<B: Block> Eq for Request<B> {}
+
+impl<B: Block> Ord for Request<B> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match (&self, &other) {
+            (Self::Block(a), Self::Block(b)) => a.cmp(b),
+            (Self::Finalized { height: a }, Self::Finalized { height: b }) => a.cmp(b),
+            (Self::Notarized { view: a }, Self::Notarized { view: b }) => a.cmp(b),
+            (Self::Block(_), Self::Finalized { .. }) => std::cmp::Ordering::Less,
+            (Self::Block(_), Self::Notarized { .. }) => std::cmp::Ordering::Less,
+            (Self::Finalized { .. }, Self::Notarized { .. }) => std::cmp::Ordering::Less,
+            (Self::Finalized { .. }, Self::Block(_)) => std::cmp::Ordering::Greater,
+            (Self::Notarized { .. }, Self::Block(_)) => std::cmp::Ordering::Greater,
+            (Self::Notarized { .. }, Self::Finalized { .. }) => std::cmp::Ordering::Greater,
+        }
```

Comment:
clarity: I would suggest adding `fn enum_value(&self) -> u8` to `Request<B>` which would return 0,1,2. Then you can first compare by `enum_value` then by `cmp` if `enum_value` is equal. You could also possibly use it in `write` but it may not be necessary